Weak runtime shape checks of matrices

sylee957 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #24149

#### Brief description of what is fixed or changed

I briefly summarize that the matrix shape check is disabled, or tested more weakly 
and left more responsibility for user to define matrix expressions in more reasonable in the first place

#### Other comments

I've also removed some logically redundant conditions for raising sympy deprecation exceptions

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- matrices
  - Allowed creating of `MatAdd`, `MatMul`, `MatPow`, `Inverse` and `HadamardProduct` with more general matrix shape. For example, previously, `MatrixSymbol('A', m, n)**2` for symbolic shape `m, n` didn't work.
<!-- END RELEASE NOTES -->

sylee957 committed:
Disable runtime type checks of matrices

oscarbenjamin commented on sympy/matrices/expressions/matpow.py:
```diff
@@ -15,8 +14,6 @@ def __new__(cls, base, exp, evaluate=False, **options):
         if not base.is_Matrix:
             raise TypeError("MatPow base should be a matrix")
-        if not base.is_square:
-            raise NonSquareMatrixError("Power of non-square matrix %s" % base)
```
Maybe `is_square` should use ternary logic and then this could be `if base.is_square is False`.

oscarbenjamin commented on sympy/matrices/expressions/tests/test_matpow.py:
```diff
@@ -104,14 +103,6 @@ def test_doit_matrix():
 
 
-def test_nonsquare():
-    A = MatrixSymbol('A', 2, 3)
-    B = ImmutableMatrix([[1, 2, 3], [4, 5, 6]])
-    for r in [-1, 0, 1, 2, S.Half, S.Pi, n]:
-        raises(NonSquareMatrixError, lambda: MatPow(A, r))
-        raises(NonSquareMatrixError, lambda: MatPow(B, r))
```
This seems like is should raise to me since we do know that the matrices are not square.

sylee957 commented:
I have different demo to use `Eq` and `And` evaluation for possibly detecting `m = m + 1` False in `doit`.
I have not tried using the equivalent logic old assumption check yet.

sylee957 commented on sympy/matrices/expressions/matadd.py:
```diff
@@ -52,11 +54,11 @@ def __new__(cls, *args, evaluate=False, check=None, _sympify=True):
+        if check:
             if not any(isinstance(i, MatrixExpr) for i in args):
                 return Add.fromiter(args)
             validate(*args)
```
The type check logic (`MatrixExpr` vs `Expr`)is inside this, so I'd give a reminder that I may need to distinguish and pull out the type checks in the future

sylee957 commented:
I have changed the shape check to `Optional[bool]` type, however, I think that there are some technical difficulties about it because of the compatibility with the existing `GenericZeroMatrix, GenericOneMatrix`.
There was also minor difficulty with the type because `MatrixBase.shape` returns `int` than `Integer`.

I've also implemented a second level of shape checks that uses more stronger `Eq` than the old assumptions check. It is able to infer $M_{n, n+1}$ is not square or such, but there can be tradeoff for speed. I've implemented it only inside `doit` logic, but I'd question if it will be necessary in the future there is another argument that if makes things slower.

Also, I think that we need to undefine `GenericOneMatrix ** -1` or should not guarantee any meaningful result for it, because it complicates the implementation.

sylee957 commented:
The benchmark results indicates that there is slowdown for `doit` when we use more stronger logic to test out the shape.

oscarbenjamin commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -171,8 +171,15 @@ def cols(self):
     @property
-    def is_square(self):
-        return self.rows == self.cols
+    def is_square(self) -> Optional[bool]:
+        try:
+            diff = self.rows - self.cols
+        except TypeError:
+            # XXX This error is raised because of the existance of GenericZeroMatrix, GenericIdentity
+            return None
+        if isinstance(diff, Basic):
+            return diff.is_zero
+        return diff == 0
```
This all seems a bit unfortunate. Ideally there would be consistent types here. I wonder if maybe the generic matrices should return some dummy symbol as well. At least for GenericIdentity it should be possible to say that `is_square` is true but maybe this is just something to be handled in those subclasses rather than here.

sylee957 replied:
This is avoided by adding `is_square` check to GenericIdentity. I'm not sure about the consequences of using dummy symbols, so they are not tried.

oscarbenjamin commented:
I was expecting at least one of these to raise:
```python
In [7]: n, m = symbols('n, m')

In [8]: M = MatrixSymbol('M', m, m)

In [9]: N = MatrixSymbol('N', n, n)

In [10]: M1 = MatrixSymbol('M1', m+1, m)

In [11]: MatAdd(M, N, M1)
Out[11]: M + N + M₁

In [12]: MatAdd(M, M1, N)
Out[12]: M + M₁ + N
```
I'm not sure what would be the best way of handling these comparisons without all to all checking.

oscarbenjamin commented:
> I was expecting at least one of these to raise:

To be clear I don't think it is completely necessary that it should raise. I'm just surprised that it does not because it least this equality evaluates:
```python
In [17]: Eq(m, m+1)
Out[17]: False
```
In general though I am not sure what is an efficient way of handling a situation where many symbolic expressions are expected to be equal without doing the `N^2` comparisons.

sylee957 commented:
> I am not sure what is an efficient way of handling a situation where many symbolic expressions are expected to be equal without doing the N^2 comparisons.

The problem is such worst time complexity is achieved every time when the input supplied by user is correct

sylee957 commented on sympy/matrices/expressions/hadamard.py:
```diff
@@ -69,15 +69,19 @@ def __new__(cls, *args, evaluate=False, check=None):
             # We currently don't have a way to support one-matrices of generic dimensions:
             raise ValueError("HadamardProduct needs at least one argument")
+
+        if not all(isinstance(arg, MatrixExpr) for arg in args):
```
I moved the check to here because it gives better type narrowing capability

sylee957 commented on sympy/matrices/expressions/tests/test_hadamard.py:
```diff
@@ -17,7 +17,6 @@
 def test_HadamardProduct():
     assert HadamardProduct(A, B, A).shape == A.shape
 
-    raises(ShapeError, lambda: HadamardProduct(A, B.T))
```
This should make sense now

sylee957 commented on sympy/matrices/expressions/tests/test_hadamard.py:
```diff
@@ -63,17 +62,13 @@ def test_hadamard():
     assert hadamard_product(A, B).doit() == hadamard_product(A, B)
-    with raises(ShapeError):
-        hadamard_product(A, C)
-        hadamard_product(A, I)
```
This test is removed because of redundancy

sylee957 commented on sympy/matrices/expressions/tests/test_inverse.py:
```diff
@@ -45,11 +45,7 @@ def test_inverse():
     assert isinstance(OneMatrix(n, n).I, Inverse)
 
 def test_inverse_non_invertible():
-    raises(NonSquareMatrixError, lambda: Inverse(A))
-    raises(NonSquareMatrixError, lambda: Inverse(A*B))
-    raises(NonSquareMatrixError, lambda: ZeroMatrix(n, m).I)
     raises(NonInvertibleMatrixError, lambda: ZeroMatrix(n, n).I)
-    raises(NonSquareMatrixError, lambda: OneMatrix(n, m).I)
```
These should make sense now

sylee957 committed:
Misc changes

sylee957 committed:
Add docstrings for new functions

sylee957 commented:
I have intermediate status report, but the benchmark indicates that I would even want to remove the shape assertion from `MatAdd, MatMul.doit` because of the performance regression of asserting symbolic matrix shape

oscarbenjamin commented:
Perhaps there could be a check only for explicit Integers.

sylee957 committed:
Remove duplicate deprecation warnings and check integer shape

sylee957 committed:
Fix flake8

sylee957 committed:
Move validators

sylee957 committed:
Add missing tests about shape error

sylee957 committed:
Move shape checking utilities to private module

sylee957 committed:
Fix tests for private methods

sylee957 commented:
I think that this should be ready for review

sylee957 committed:
Test more faster square check

sylee957 commented:
Okay now when the square check is finally restricted to integers, the negative benchmark results are entirely gone.

oscarbenjamin commented on sympy/matrices/expressions/tests/test_special.py:
```diff
@@ -80,7 +79,8 @@ def test_generic_identity():
     assert I.is_Identity
-    assert I**-1 == I
+    # XXX Should we need to undefine this?
+    # assert I**-1 == I
```
Why is this changed?

oscarbenjamin commented:
I had one question but otherwise this looks good to me.

sylee957 committed:
Uncomment test

asmeurer commented:
It's good to see this addressed. It might still be useful to have a method that does a strict shape check, like what used to happen automatically.
